File: src/server/lib/entry-windows.ts
/**
 * entry-windows.ts — W23H.3 allowed_entry_windows parser + checker
 *
 * Parses window specs of the form "HH:MM-HH:MM TZ" and checks whether a
 * bar timestamp falls inside any configured window.
 *
 * Boundary semantics: [start, end) — left-inclusive, right-exclusive.
 * So "09:45-12:00 ET" includes 09:45 but NOT 12:00. This avoids overlap
 * when windows abut ("09:45-12:00 ET" + "12:00-14:00 ET" have no gap).
 *
 * Timezone shorthands: "ET" → "America/New_York", "PT" → "America/Los_Angeles",
 * "CT" → "America/Chicago", "MT" → "America/Denver", "UTC" → "UTC".
 * Any other string is passed through as an IANA name.
 *
 * This module is used in three places:
 *   1. paper-signal-service.ts: signal-time enforcement
 *   2. framework-overlay.ts: config field definition (type-only)
 *   3. Tests
 *
 * Python mirror: src/engine/entry_windows.py
 */

// ─── Timezone shorthands → IANA names ───────────────────────────────────────

const TZ_SHORTHANDS: Record<string, string> = {
  ET: "America/New_York",
  EST: "America/New_York",
  EDT: "America/New_York",
  PT: "America/Los_Angeles",
  PST: "America/Los_Angeles",
  PDT: "America/Los_Angeles",
  CT: "America/Chicago",
  CST: "America/Chicago",
  CDT: "America/Chicago",
  MT: "America/Denver",
  MST: "America/Denver",
  MDT: "America/Denver",
  UTC: "UTC",
};

function resolveTimezone(tz: string): string {
  return TZ_SHORTHANDS[tz.toUpperCase()] ?? tz;
}

// ─── Types ───────────────────────────────────────────────────────────────────

export interface EntryWindow {
  /** Minutes since 00:00 in window timezone. E.g. 9*60+45 = 585 for 09:45. */
  startMinutesOfDay: number;
  /** Minutes since 00:00 in window timezone. Right-exclusive. */
  endMinutesOfDay: number;
  /** IANA timezone name e.g. "America/New_York". */
  timezone: string;
  /** Original spec string, for diagnostics. */
  spec: string;
}

// ─── Parser ─────────────────────────────────────────────────────────────────

/**
 * Parse a window spec string: "HH:MM-HH:MM TZ"
 * Throws on malformed input — validation happens at config time, not signal time.
 *
 * Examples:
 *   parseEntryWindow("09:45-12:00 ET")   → {start:585, end:720, tz:"America/New_York"}
 *   parseEntryWindow("13:30-15:30 ET")   → {start:810, end:930, tz:"America/New_York"}
 *   parseEntryWindow("09:00-17:00 UTC")  → {start:540, end:1020, tz:"UTC"}
 */
export function parseEntryWindow(spec: string): EntryWindow {
  if (typeof spec !== "string" || spec.trim() === "") {
    throw new Error(`parseEntryWindow: empty or non-string spec: ${JSON.stringify(spec)}`);
  }

  // Split at first space — left side is time range, right side is TZ
  const trimmed = spec.trim();
  const spaceIdx = trimmed.lastIndexOf(" ");
  if (spaceIdx < 0) {
    throw new Error(`parseEntryWindow: missing timezone in spec "${spec}" — expected "HH:MM-HH:MM TZ"`);
  }

  const timeRange = trimmed.slice(0, spaceIdx).trim();
  const tzRaw = trimmed.slice(spaceIdx + 1).trim();

  if (!tzRaw) {
    throw new Error(`parseEntryWindow: missing timezone in spec "${spec}"`);
  }

  // Parse "HH:MM-HH:MM"
  const dashIdx = timeRange.indexOf("-");
  if (dashIdx < 0) {
    throw new Error(`parseEntryWindow: missing '-' separator in time range "${timeRange}" (spec: "${spec}")`);
  }

  const startStr = timeRange.slice(0, dashIdx);
  const endStr = timeRange.slice(dashIdx + 1);

  function parseTime(t: string, label: string): number {
    const parts = t.split(":");
    if (parts.length !== 2) {
      throw new Error(`parseEntryWindow: ${label} "${t}" must be HH:MM (spec: "${spec}")`);
    }
    const h = parseInt(parts[0], 10);
    const m = parseInt(parts[1], 10);
    if (isNaN(h) || isNaN(m)) {
      throw new Error(`parseEntryWindow: ${label} "${t}" contains non-numeric value (spec: "${spec}")`);
    }
    if (h < 0 || h > 23) {
      throw new Error(`parseEntryWindow: ${label} hour ${h} out of range [0,23] (spec: "${spec}")`);
    }
    if (m < 0 || m > 59) {
      throw new Error(`parseEntryWindow: ${label} minute ${m} out of range [0,59] (spec: "${spec}")`);
    }
    return h * 60 + m;
  }

  const startMinutesOfDay = parseTime(startStr, "start");
  const endMinutesOfDay = parseTime(endStr, "end");

  if (endMinutesOfDay <= startMinutesOfDay) {
    throw new Error(
      `parseEntryWindow: end (${endStr}) must be after start (${startStr}) — overnight windows not supported (spec: "${spec}")`,
    );
  }

  const timezone = resolveTimezone(tzRaw);

  // Validate that the IANA timezone resolves without throwing at runtime.
  // Intl.DateTimeFormat constructor throws on bad timezone names.
  try {
    new Intl.DateTimeFormat("en-US", { timeZone: timezone });
  } catch {
    throw new Error(`parseEntryWindow: unrecognized timezone "${tzRaw}" (resolved: "${timezone}") in spec "${spec}"`);
  }

  return { startMinutesOfDay, endMinutesOfDay, timezone, spec };
}

/**
 * Parse an array of window specs. Throws on the first malformed spec.
 * Returns [] when input is undefined or empty.
 */
export function parseEntryWindows(specs: string[] | undefined | null): EntryWindow[] {
  if (!specs || specs.length === 0) return [];
  return specs.map(parseEntryWindow);
}

// ─── Checker ─────────────────────────────────────────────────────────────────

/**
 * Return the "minutes of day" for a UTC timestamp converted to a given timezone.
 * Uses Intl.DateTimeFormat to handle DST correctly.
 */
function toMinutesOfDayInTz(barTsUtc: Date, timezone: string): number {
  // Intl gives us the wall-clock time in the target timezone.
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: timezone,
    hour: "numeric",
    minute: "numeric",
    hour12: false,
  }).formatToParts(barTsUtc);

  let h = 0;
  let m = 0;
  for (const part of parts) {
    if (part.type === "hour") {
      h = parseInt(part.value, 10);
      // hour12:false returns "24" for midnight — normalize to 0
      if (h === 24) h = 0;
    } else if (part.type === "minute") {
      m = parseInt(part.value, 10);
    }
  }
  return h * 60 + m;
}

/**
 * Check if barTsUtc falls within a single entry window.
 * Boundary: [start, end) — start-inclusive, end-exclusive.
 */
export function isBarInWindow(barTsUtc: Date, window: EntryWindow): boolean {
  const minutesOfDay = toMinutesOfDayInTz(barTsUtc, window.timezone);
  return minutesOfDay >= window.startMinutesOfDay && minutesOfDay < window.endMinutesOfDay;
}

/**
 * Check if barTsUtc falls within ANY of the configured entry windows.
 * Returns false when windows array is empty (no restriction defined — callers
 * must handle the empty-array case as "no restriction" themselves, since this
 * function returns false for an empty array to distinguish from "in a window").
 *
 * Callers should check: windows.length === 0 || isBarInAnyWindow(ts, windows)
 */
export function isBarInAnyWindow(barTsUtc: Date, windows: EntryWindow[]): boolean {
  if (windows.length === 0) return false;
  return windows.some((w) => isBarInWindow(barTsUtc, w));
}

File: src/engine/entry_windows.py
"""
entry_windows.py — W23H.3 allowed_entry_windows parser + checker (Python mirror)

Parses window specs of the form "HH:MM-HH:MM TZ" and checks whether a
bar timestamp falls inside any configured window.

Boundary semantics: [start, end) — left-inclusive, right-exclusive.
So "09:45-12:00 ET" includes 09:45:00 but NOT 12:00:00.

Timezone shorthands:
  ET → America/New_York
  PT → America/Los_Angeles
  CT → America/Chicago
  MT → America/Denver
  UTC → UTC
  Any other string is passed through as an IANA name.

TypeScript mirror: src/server/lib/entry-windows.ts
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime, timezone, timedelta
from typing import Optional, Sequence

try:
    from zoneinfo import ZoneInfo, ZoneInfoNotFoundError  # Python 3.9+
except ImportError:  # pragma: no cover
    from backports.zoneinfo import ZoneInfo, ZoneInfoNotFoundError  # type: ignore

# ─── Timezone shorthands ─────────────────────────────────────────────────────

TZ_SHORTHANDS: dict[str, str] = {
    "ET": "America/New_York",
    "EST": "America/New_York",
    "EDT": "America/New_York",
    "PT": "America/Los_Angeles",
    "PST": "America/Los_Angeles",
    "PDT": "America/Los_Angeles",
    "CT": "America/Chicago",
    "CST": "America/Chicago",
    "CDT": "America/Chicago",
    "MT": "America/Denver",
    "MST": "America/Denver",
    "MDT": "America/Denver",
    "UTC": "UTC",
}


def resolve_timezone(tz_raw: str) -> str:
    """Resolve a shorthand or IANA timezone name to a canonical IANA name."""
    return TZ_SHORTHANDS.get(tz_raw.upper(), tz_raw)


# ─── Types ───────────────────────────────────────────────────────────────────

@dataclass(frozen=True)
class EntryWindow:
    """A parsed entry window.

    start_minutes_of_day and end_minutes_of_day are minutes-since-midnight in
    the window's timezone. Boundary: [start, end) left-inclusive, right-exclusive.
    """
    start_minutes_of_day: int
    end_minutes_of_day: int
    timezone: str       # IANA name
    spec: str           # original spec string, for diagnostics


# ─── Parser ──────────────────────────────────────────────────────────────────

_TIME_RE = re.compile(r"^(\d{2}):(\d{2})-(\d{2}):(\d{2})$")


def parse_entry_window(spec: str) -> EntryWindow:
    """Parse "HH:MM-HH:MM TZ" → EntryWindow. Raises ValueError on malformed input.

    Examples::

        parse_entry_window("09:45-12:00 ET")
        # EntryWindow(start_minutes_of_day=585, end_minutes_of_day=720,
        #             timezone="America/New_York", spec="09:45-12:00 ET")

        parse_entry_window("13:30-15:30 ET")
        # EntryWindow(start_minutes_of_day=810, end_minutes_of_day=930, ...)
    """
    if not isinstance(spec, str) or not spec.strip():
        raise ValueError(f"parse_entry_window: empty or non-string spec: {spec!r}")

    trimmed = spec.strip()
    # Split at last whitespace — left side is time range, right side is TZ
    parts = trimmed.rsplit(" ", 1)
    if len(parts) != 2:
        raise ValueError(
            f'parse_entry_window: missing timezone in spec "{spec}" '
            f'— expected "HH:MM-HH:MM TZ"'
        )

    time_range, tz_raw = parts[0].strip(), parts[1].strip()
    if not tz_raw:
        raise ValueError(f'parse_entry_window: missing timezone in spec "{spec}"')

    m = _TIME_RE.match(time_range)
    if not m:
        raise ValueError(
            f'parse_entry_window: time range "{time_range}" must be HH:MM-HH:MM '
            f'(spec: "{spec}")'
        )

    sh, sm, eh, em = int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))

    if sm > 59:
        raise ValueError(
            f'parse_entry_window: start minute {sm} out of range [0,59] (spec: "{spec}")'
        )
    if em > 59:
        raise ValueError(
            f'parse_entry_window: end minute {em} out of range [0,59] (spec: "{spec}")'
        )
    if sh > 23:
        raise ValueError(
            f'parse_entry_window: start hour {sh} out of range [0,23] (spec: "{spec}")'
        )
    if eh > 23:
        raise ValueError(
            f'parse_entry_window: end hour {eh} out of range [0,23] (spec: "{spec}")'
        )

    start_min = sh * 60 + sm
    end_min = eh * 60 + em

    if end_min <= start_min:
        raise ValueError(
            f'parse_entry_window: end ({m.group(3)}:{m.group(4)}) must be after '
            f'start ({m.group(1)}:{m.group(2)}) — overnight windows not supported '
            f'(spec: "{spec}")'
        )

    iana_tz = resolve_timezone(tz_raw)

    # Validate the IANA timezone name by constructing a ZoneInfo object.
    try:
        ZoneInfo(iana_tz)
    except (ZoneInfoNotFoundError, KeyError) as exc:
        raise ValueError(
            f'parse_entry_window: unrecognized timezone "{tz_raw}" '
            f'(resolved: "{iana_tz}") in spec "{spec}": {exc}'
        ) from exc

    return EntryWindow(
        start_minutes_of_day=start_min,
        end_minutes_of_day=end_min,
        timezone=iana_tz,
        spec=spec,
    )


def parse_entry_windows(specs: Optional[Sequence[str]]) -> list[EntryWindow]:
    """Parse a list of window specs. Raises on first malformed spec. Returns [] for None/empty."""
    if not specs:
        return []
    return [parse_entry_window(s) for s in specs]


# ─── Checker ─────────────────────────────────────────────────────────────────

def to_minutes_of_day_in_tz(ts_utc: datetime, iana_tz: str) -> int:
    """Convert a UTC datetime to minutes-of-day in the given IANA timezone.

    Handles DST correctly because ZoneInfo applies the correct UTC offset for
    the moment represented by ts_utc.
    """
    if ts_utc.tzinfo is None:
        ts_utc = ts_utc.replace(tzinfo=timezone.utc)
    local_dt = ts_utc.astimezone(ZoneInfo(iana_tz))
    return local_dt.hour * 60 + local_dt.minute


def is_bar_in_window(ts_utc: datetime, window: EntryWindow) -> bool:
    """Return True if ts_utc falls in [window.start, window.end) in window.timezone."""
    minutes = to_minutes_of_day_in_tz(ts_utc, window.timezone)
    return window.start_minutes_of_day <= minutes < window.end_minutes_of_day


def is_bar_in_any_window(ts_utc: datetime, windows: list[EntryWindow]) -> bool:
    """Return True if ts_utc falls in ANY of the windows.

    Returns False for empty windows list — callers are responsible for the
    semantics: ``not windows or is_bar_in_any_window(ts, windows)`` for
    "no restriction when list is empty".
    """
    if not windows:
        return False
    return any(is_bar_in_window(ts_utc, w) for w in windows)


# ─── Pine helper ─────────────────────────────────────────────────────────────

def window_to_pine_time_string(window: EntryWindow) -> str:
    """Convert a parsed EntryWindow to a Pine-compatible time() argument string.

    Pine time() format: "HHMM-HHMM" (no colon).
    Example: start=585 (09:45), end=720 (12:00) → "0945-1200"
    """
    sh, sm = divmod(window.start_minutes_of_day, 60)
    eh, em = divmod(window.end_minutes_of_day, 60)
    return f"{sh:02d}{sm:02d}-{eh:02d}{em:02d}"
